### PR TITLE
feat(batch): support `mysql_query` for mysql batch ingestion

### PR DESCRIPTION
**Pull request by kwannoel**
I hereby agree to the terms of the [RisingWave Labs, Inc. Contributor License Agreement](https://gist.github.com/TennyZhuang/f00be7f16996ea48effb049aa7be4d66#file-rw_cla).

## What's changed and what's your intention?

We support `mysql_query` in this PR. We extract common deserialization logic between mysql cdc parser and mysql_query (introduced in this PR). We also handle serde for `bit`. `bit(1)` will just be treat as `boolean`. `bit(>1)` will be treated as `bytea`.

I also added a `local-inline-source-test` profile, which complements `ci-inline-source-test`, so we can easily run source tests locally.

You may consult the table in the release notes for the column type mapping for `MySQL -> RisingWave` after this PR.

## Checklist

- [ ] I have written necessary rustdoc comments
- [ ] I have added necessary unit tests and integration tests
- [ ] I have added test labels as necessary. See [details](https://github.com/risingwavelabs/risingwave/blob/main/docs/developer-guide.md#ci-labels-guide).
- [ ] I have added fuzzing tests or opened an issue to track them. (Optional, recommended for new SQL features #7934).
- [ ] My PR contains breaking changes. (If it deprecates some features, please create a tracking issue to remove them in the future).
- [ ] All checks passed in `./risedev check` (or alias, `./risedev c`)
- [ ] My PR changes performance-critical code. (Please run macro/micro-benchmarks and show the results.)

- [ ] My PR contains critical fixes that are necessary to be merged into the latest release. (Please check out the [details](https://github.com/risingwavelabs/risingwave/blob/main/CONTRIBUTING.md))

## Documentation

- [ ] My PR needs documentation updates. (Please use the **Release note** section below to summarize the impact on users)

## Release note

This feature should be a TECHNICAL PREVIEW. See below.

This PR introduces the `mysql_query` tvf. Its syntax will very likely change down the road. So the interface and examples you see below will be obsolete in subsequent versions. That being said the core functionality will still exist.

Type mapping table. The only changes are for `bit(1)` and `bit(>1)`. Previously both of these map to `NULL`. Now these are supported.

Bool and boolean are mapped to `smallint` as well, because they are actually an alias of `tinyint(1)`. If we choose to map them to `boolean`, it means if someone declared a `tinyint(1)` intentionally, it will be converted to `boolean` as well.

Instead we just opt for a more consistent approach: all `tinyint`, including `(bool/boolean)` will map to `smallint`, users may convert `bit(1)` into RW boolean if they wish to instead.

| MySQL Type | RisingWave Type |
| - | - |
| bit(1) | boolean |
| bit(>1) | bytea |
| bool/boolean | smallint |
| tinyint | smallint |
| smallint | smallint |
| mediumint | int |
| int | int |
| bigint | bigint |
| float | float32 |
| double | float64 |
| decimal | decimal |
| numeric | decimal |
| year | int |
| date | date |
| time | time |
| datetime | timestamp |
| timestamp | timestamptz |
| varchar | varchar |
| char | varchar |
| json | jsonb |
| blob | bytea |
| tinyblob | bytea |
| mediumblob | bytea |
| longblob | bytea |
| array | unsupported |
| enum | unsupported |
| set | unsupported |
| geometry | unsupported |
| null | unsupported |

After this PR we support `mysql_query`.

Syntax:
```
mysql_query(hostname varchar, port varchar, username varchar, password varchar, database_name varchar, query varchar)
```

### Example

Mysql:
```sql
CREATE TABLE test (
    id bigint primary key,
    v0 bit,
    v1 bool,
    v2 tinyint(1),
    v3 tinyint(2),
    v4 smallint,
    v5 mediumint,
    v6 integer,
    v7 bigint,
    v8 float,
    v9 double,
    v10 numeric(4, 2),
    v11 decimal(4, 2),
    v12 char(255),
    v13 varchar(255),
    v14 bit(10),
    v15 tinyblob,
    v16 blob,
    v17 mediumblob,
    v18 longblob,
    v19 date,
    v20 time,
    v21 timestamp,
    v22 json
);
INSERT INTO test SELECT
  1 as id,
  true as v0,
  true as v1,
  2 as v2,
  3 as v3,
  4 as v4,
  5 as v5,
  6 as v6,
  7 as v7,
  1.08 as v8,
  1.09 as v9,
  1.10 as v10,
  1.11 as v11,
  'char' as v12,
  'varchar' as v13,
  b'1010' as v14,
  x'16' as v15,
  x'17' as v16,
  x'18' as v17,
  x'19' as v18,
  '2021-01-01' as v19,
  '12:34:56' as v20,
  '2021-01-01 12:34:56' as v21,
  JSON_OBJECT('key1', 1, 'key2', 'abc');
```

rw:
```sql
select * from mysql_query('$MYSQL_HOST', '$MYSQL_TCP_PORT', '$RISEDEV_MYSQL_USER', '$MYSQL_PWD', 'tvf', 'select * from test;');
----
1 t 1 2 3 4 5 6 7 1.08 1.09 1.10 1.11 char varchar \x000a \x16 \x17 \x18 \x19 2021-01-01 12:34:56 2021-01-01 12:34:56+00:00 {"key1": 1, "key2": "abc"}
```

### CODE REVIEW

File: src/batch/src/executor/mysql_query.rs
Comment (fuyufjh) [inline src/batch/src/executor/mysql_query.rs:1]: cc @StrikeW to see whether there some duplications

File: src/batch/src/executor/mysql_query.rs [hunk 1]
```diff
@@ -0,0 +1,166 @@
...
+        let query = self.query;
+        let mut query_iter = conn
+            .query_iter(query)
+            .await
+            .context("failed to execute my_sql_query in batch executor")?;
+        let Some(row_stream) = query_iter.stream::<mysql_async::Row>().await? else {
+            bail!("failed to get row stream from mysql query")
+        };
+
+        let mut builder = DataChunkBuilder::new(self.schema.data_types(), 1024);
```
Comment (chenzl25) [inline src/batch/src/executor/mysql_query.rs:124]: We can use `batch_chunk_size` instead of 1024.

File: src/frontend/src/expr/table_function.rs [hunk 2]
```diff
@@ -350,6 +353,157 @@ impl TableFunction {
...
+                    }
+                }
+            }
+            evaled_args
+        };
+
+        #[cfg(madsim)]
+        {
+            return Err(crate::error::ErrorCode::BindError(
+                "postgres_query can't be used in the madsim mode".to_string(),
```
Comment (chenzl25) [inline src/frontend/src/expr/table_function.rs:399]: `mysql_query`

File: src/frontend/src/expr/table_function.rs [hunk 3]
```diff
@@ -350,6 +353,157 @@ impl TableFunction {
...
+            .into());
+        }
+
+        #[cfg(not(madsim))]
+        {
+            let schema = tokio::task::block_in_place(|| {
+                RUNTIME.block_on(async {
+                    let database_opts: mysql_async::Opts = mysql_async::OptsBuilder::default()
+                        .ip_or_hostname(evaled_args[0].clone())
+                        .tcp_port(evaled_args[1].parse::<u16>().unwrap())
```
Comment (chenzl25) [inline src/frontend/src/expr/table_function.rs:410]: It'd be better to check the parse result and throw an error instead of `unwrap()`.

**Review by chenzl25 — APPROVED** [review]
Rest LGTM.

**Review by StrikeW — APPROVED** [review]
LGTM

File: src/frontend/src/optimizer/logical_optimization.rs [hunk 4]
```diff
@@ -136,6 +136,8 @@ static TABLE_FUNCTION_CONVERT: LazyLock<OptimizationStage> = LazyLock::new(|| {
             TableFunctionToFileScanRule::create(),
             // Apply postgres query rule next
             TableFunctionToPostgresQueryRule::create(),
+            // Apply mysql query rule next
+            TableFunctionToMySqlQueryRule::create(),
```
[the remark this answers is not in the document]
Comment (kwannoel) [reply src/frontend/src/optimizer/logical_optimization.rs:140]: Wait for `CONNECTION` first: https://github.com/risingwavelabs/risingwave/issues/19222.

We can refactor pg and mysql then.

File: src/common/src/mysql.rs [hunk 5]
```diff
@@ -0,0 +1,191 @@
...
+        }
+    }};
+}
+
+/// The decoding result can be interpreted as follows:
+/// Ok(value) => The value was found and successfully decoded.
+/// Err(error) => The value was found but could not be decoded,
+///               either because it was not supported,
+///               or there was an error during conversion.
+pub fn mysql_datum_to_rw_datum(
```
[the remark this answers is not in the document]
Comment (kwannoel) [reply src/common/src/mysql.rs]: Fixed.

File: src/frontend/src/expr/table_function.rs (hunk 3 above)
Comment (kwannoel) [reply src/frontend/src/expr/table_function.rs]: Fixed

File: e2e_test/source_legacy/cdc_inline/auto_schema_map_mysql.slt [hunk 6]
```diff
@@ -174,8 +174,8 @@ SELECT
...
-1 NULL -128 -32767 -8388608 -2147483647 -9223372036854775807 -10 -10000 -10000 a b \x000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000
+0 f NULL NULL -8388608 -2147483647 9223372036854775806 -10 -10000 -10000 c d \x000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000
```
Comment (xxchan) [inline e2e_test/source_legacy/cdc_inline/auto_schema_map_mysql.slt:177]: why is this changed? Could you please note the "changes" of the "Type mapping table" in the release note?

**Review by xxchan — APPROVED** [review]

File: e2e_test/source_legacy/cdc_inline/auto_schema_map_mysql.slt (hunk 6 above)
Comment (kwannoel) [reply e2e_test/source_legacy/cdc_inline/auto_schema_map_mysql.slt:177]: Updated the release notes, previously only mentioned them in the PR description. It's because we support `bit` conversion now.
